contracts/atm/ATMLiquidityMining.sol

```diff
@@ -0,0 +1,332 @@
+        @param tlrTokenProxy TLR Token instance associated with this ATM Liquidity Mining program.
+     */
+    function initialize(
+        address settingsAddress,
+        address atmGovernanceProxy, 
+        address tlrTokenProxy
+    ) 
+        external
+        isNotInitialized()
+    {
+        _setSettings(settingsAddress);
+        governance = ATMGovernanceInterface(atmGovernanceProxy);
+        tlrToken = TLRTokenInterface(tlrTokenProxy);
+        TInitializable._initialize();
+    }
+
+    /**
+        @notice Adds a new address to the blacklisted list. This address will no longer be able to 
+        stake on this ATM Liquidity Mining program intance. 
+        @param notAllowed address we will blacklist.
+     */
+    function addNotAllowedAddress(address notAllowed)
+        external
+        onlyPauser()
+        isInitialized()
+    {
+        notAllowedAddresses[notAllowed] = true;
+        emit NotAllowedAddressAdded(msg.sender, notAllowed);
+    }
+   
+    /**
+        @notice Removes an address from the blacklisted list. This address will now start
+        being able to stake on this ATM Liquidity Mining program intance. 
+        @param notAllowed address we will blacklist.
+     */
+    function removeNotAllowedAddress(address notAllowed)
+        external
+        onlyPauser()
+        isInitialized()
+    {
+        notAllowedAddresses[notAllowed] = false;
+        emit NotAllowedAddressRemoved(msg.sender, notAllowed);
+    }
+    
+    /**
+        @notice End users stake their own tTokens on this ATM Liquidity Mining program to 
+            earn TLR. This operation updates the user stake info for the sender (userStakeInfo[msg.sender]).
+        @param tToken address of the tToken they will stake.
+        @param amount amount of tToken to stake.
+     */
+    function stake(address tToken, uint256 amount) 
+        external 
+        onlyAllowed()
+        isInitialized() 
+        whenNotPaused()
+    {
+        // TODO: Check tToken is a Teller whitelisted token on settings().
+        //require(settings().tTokensRegistry().istTokenValid(tToken), "TTOKEN_IS_NOT_REGISTERED");
+        require(amount > 0, "STAKING_ZERO_NOT_ALLOWED");
+        // Checking tToken balance
+        require(TToken(tToken).balanceOf(msg.sender) >= amount, "INSUFFICIENT_TTOKENS_TO_STAKE");
+        // Transferring tTokens for staking
+        require(TToken(tToken).transferFrom(msg.sender, address(this), amount), "STAKE_TTOKEN_TRANSFER_FAILED");
+        
+        // Update use stake info
+        ATMLibrary.UserStakeInfo memory userInfo = ATMLibrary.UserStakeInfo ( {
+                lastRewardedBlock: block.number,
+                tTokenStakedBalance: _incrementTTokenBalance(tToken, amount), // Staking
+                accruedTLRBalance: _getTLRTotalBalance(tToken)
+            });
+        userStakeInfo[msg.sender][tToken] = userInfo;
```

Comment:
Given Guille's suggestion before, this could simplify it and remove the need for the `_incrementTTokenBalance` function.

```
           TMLibrary.UserStakeInfo memory userInfo = userStakeInfo[msg.sender[tToken];
           userInfo.lastRewardedBlock: block.number;
           userInfo.tTokenStakedBalance = userInfo.tTokenStakedBalance.add(amount);
           userInfo.accruedTLRBalance  _getTLRTotalBalance(tToken);
```

Reply:
I used that notation to make it easier to read that first we create the struct with all it's data and then we assign the entire struct to the mapping. Readability purpose only.

Reply:
I would argue that the readability is better the other way. 
With mappings, we do not need to instantiate values, but rather they are always defaulted to empty values. Therefore, all we need to do is read the value in the mapping and update it. I would also argue that this saves gas as we are not creating a new `UserStakeInfo` struct and then overwriting the existing struct in the mapping. Rather we are just updating values in the existing struct.

Reply:
I created a new version please check this again. Basically I will calculate accrued TLR in one step, then receive transferred tokens, then assign state variable. Please check the new version and let me know your thoughts.